components/docker-auth/cmd/authn/authentication.go

```diff
@@ -0,0 +1,113 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/cellery-io/cellery-hub/components/docker-auth/pkg/extension"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/golang/glog"
+)
+
+const (
+	issuerClaim           = "iss"
+	subjectClaim          = "sub"
+	authTokenIssuerEnvVar = "REGISTRY_AUTH_TOKEN_ISSUER"
+	idpCertEnvVar         = "IDP_CERT"
+	dockerAuthCertEnvVar  = "REGISTRY_AUTH_TOKEN_ROOTCERTBUNDLE"
+)
+
+func readCerts(certPath string) ([]byte, error) {
+	key, err := ioutil.ReadFile(os.Getenv(certPath))
+	if err != nil {
+		return nil, err
+	}
+	return key, nil
+}
+
+func getJWTClaim(token interface{}) jwt.MapClaims {
+	jwtToken, _ := jwt.Parse(token.(string), nil)
+	claims := jwtToken.Claims.(jwt.MapClaims)
+	return claims
+}
+
+func validateToken(inToken interface{}, cert []byte) (bool, error) {
+	publicRSA, err := jwt.ParseRSAPublicKeyFromPEM(cert)
+	if err != nil {
+		return false, err
+	}
+	token, err := jwt.Parse(inToken.(string), func(token *jwt.Token) (interface{}, error) {
+		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
+			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
+		}
+		return publicRSA, err
+	})
+	if token != nil {
+		if token.Valid {
+			return true, nil
+		} else {
+			return false, err
+		}
+	}
+	return false, err
+}
+
+func main() {
+	var certificateInUse []byte
+	flag.Parse()
+	text := extension.ReadStdIn()
+	glog.Infof("text received from CLI : %s", text)
+	credentials := strings.Split(text, " ")
+
+	if len(credentials) != 2 {
+		glog.Info("Received more than two parameters")
+		glog.Flush()
+		os.Exit(extension.ErrorExitCode)
+	}
+	uName := credentials[0]
+	token := credentials[1]
+
+	claim := getJWTClaim(token)
+	iss := claim[issuerClaim].(string)
+	sub := claim[subjectClaim].(string)
+
+	glog.Info("Token issuer :" + iss)
+	glog.Info("Subject :" + sub)
+
+	if sub != uName {
+		glog.Info("Username does not match with subject in JWT")
+		glog.Flush()
+		os.Exit(extension.ErrorExitCode)
+	}
+
+	certificateInUse, err := readCerts(idpCertEnvVar)
+
+	if iss == authTokenIssuerEnvVar {
+		certificateInUse, err = readCerts(dockerAuthCertEnvVar)
+	}
+	if err != nil {
+		glog.Info(err)
```

Comment:
This should be an error log and would be best to add our own message with more context